src/batch.ts

```diff
@@ -0,0 +1,253 @@
+                               logarithmic: boolean,
+                               min: number, max: number): BatchPars {
+    const value = getParameterValueAsNumber(base, name);
+    if (min > value) {
+        throw Error(`Expected lower bound to be no greater than ${value}`);
+    }
+    if (max < value) {
+        throw Error(`Expected upper bound to be no less than ${value}`);
+    }
+    if (min >= max) {
+        throw Error("Expected upper bound to be greater than lower bound");
+    }
+    if (count < 2) {
+        throw Error("Must include at least 2 traces in the batch");
+    }
+    const values = logarithmic ?
+        gridLog(min, max, count) : grid(min, max, count);
+    return {base, name, values};
+}
+
+/**
+ * @param base The base set of parameters
+ *
+ * @param name Name of the parameter to change
+ *
+ * @param count The number of simulations to run
+ *
+ * @param logarithmic If `true`, the points are spaced on a
+ * logarithmic scale rather than arithmetic.
+ *
+ * @param displace The *percentage* amount to displace the current
+ * value of the parameter `name` (so a value of 15 creates values
+ * ranging from 15% below to 15% above the current value).
+ */
+export function batchParsDisplace(base: UserType, name: string, count: number,
+                                  logarithmic: boolean,
+                                  displace: number): BatchPars {
+    const value = getParameterValueAsNumber(base, name);
+    const delta = displace / 100;
+    const min = value * (1 - delta);
+    const max = value * (1 + delta);
+    return batchParsRange(base, name, count, logarithmic, min, max);
+}
+
+export function updatePars(base: UserType, name: string, value: number) {
+    const ret = new Map(base);
+    ret.set(name, value);
+    return ret;
+}
+
+function getParameterValueAsNumber(pars: UserType, name: string): number {
+    const value = pars.get(name);
+    if (value === undefined) {
+        throw Error(`Expected a value for '${name}'`);
+    }
+    if (typeof value !== "number") {
+        throw Error(`Expected a number for '${name}'`);
+    }
+    return value;
+}
+
+/**
+ * Collection of extreme values; this is a templated type
+ */
+export interface Extremes<T> {
+    /** The time that a series reached its minimum */
+    tMin: T;
+    /** The time that a series reached its maximum */
+    tMax: T;
+    /** The minimum value of a series */
+    yMin: T;
+    /** The maximum value of a series */
+    yMax: T;
+}
+
+// Later, we might do some polishing of these, which should make it
+// both faster and more accurate, but we'll need to pass in the
+// correct interpolating function too.
+function findExtremes(s: Series): Extremes<number> {
+    const idxMin = whichMin(s.y);
+    const idxMax = whichMax(s.y);
+    const tMin = s.x[idxMin];
+    const tMax = s.x[idxMax];
+    const yMin = s.y[idxMin];
+    const yMax = s.y[idxMax];
+    return {tMax, tMin, yMax, yMin};
+}
```

Comment:
Does this mean that it's always finding all the extremes? But when the client wants to change e.g. from valueAtMin to timeAtValueMax, it will call the `extreme` method again, and it will all get recalculated? Could only calculate the extreme you actually need this time? Maybe so comparativley cheap that it's not worth it!

Reply:
This is cached - https://github.com/mrc-ide/odin-js/blob/mrc-3393/src/batch.ts#L80

So the first time that any extreme value is requested, all (both) are calculated (yMin and tMin are found in the same search, as are yMax/tMax). The reason for doing both at the same time is because they all share the same broad grid of values over the time domain.

Once we move to actually optimising the solutions a slightly more clever approach might be needed.

Reply:
Ah, sorry, missed that!